src/ansi-c/c_typecheck_code.cpp

```diff
@@ -903,39 +922,91 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
     throw 0;
 }
 
+void c_typecheck_baset::typecheck_conditional_targets(
+  exprt::operandst &targets,
+  const std::function<void(exprt &)> typecheck_target,
+  const std::string &clause_type)
+{
+  exprt::operandst tmp;
+  bool must_throw = false;
+
+  for(auto &target : targets)
+  {
+    if(!can_cast_expr<conditional_target_group_exprt>(target))
+    {
+      must_throw = true;
+      error().source_location = target.source_location();
+      error() << "expected ID_conditional_target_group expression in " +
+                   clause_type + "clause, found "
+              << id2string(target.id()) << eom;
+    }
+
+    auto &conditional_target_group = to_conditional_target_group_expr(target);
+    validate_expr(conditional_target_group);
+
+    // typecheck condition
+    auto &condition = conditional_target_group.condition();
+    typecheck_spec_condition(condition);
+
+    if(condition.is_true())
+    {
+      // if the condition is trivially true,
+      // simplify expr and expose the bare expressions
+      for(auto &actual_target : conditional_target_group.targets())
+      {
+        typecheck_target(actual_target);
+        tmp.push_back(actual_target);
+      }
+    }
+    else
+    {
+      // if the condition is not trivially true, typecheck in place
+      for(auto &actual_target : conditional_target_group.targets())
+      {
+        typecheck_target(actual_target);
+      }
+      tmp.push_back(std::move(target));
+    }
+  }
+
+  // at least one target was not as expected
+  if(must_throw)
+    throw 0;
+
+  // now each target is either:
+  // - a simple side-effect-free unconditional lvalue expression or
+  // - a conditional target group expression with a non-trivial condition
+
+  // update original vector in-place
+  std::swap(targets, tmp);
+}
+
+void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
+{
+  const std::function<void(exprt &)> typecheck_target = [&](exprt &target) {
+    typecheck_spec_assigns_target(target);
+  };
+  typecheck_conditional_targets(targets, typecheck_target, "assigns");
+}
+
 void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
 {
   // compute type
   typecheck_expr(target);
 
   // fatal errors
-  if(target.type().id() == ID_empty)
+  bool is_empty_type = target.type().id() == ID_empty;
+  bool is_assignable_typedef =
+    target.type().get(ID_C_typedef) == CPROVER_PREFIX "assignable_t";
+  // only allow void type if it is the typedef CPROVER_PREFIX "assignable_t"
+  if(target.type().id() == ID_empty && !is_assignable_typedef)
   {
     error().source_location = target.source_location();
     error() << "void-typed expressions not allowed as assigns clause targets"
             << eom;
     throw 0;
   }
```

Comment:
What if we actually allowed `void`-typed expressions? Would there really be anything wrong in doing that? Wouldn't that allow us to get rid of `__CPROVER_assignable_t`?

Reply:
The type is meant to documents the fact that the function specifies sets of assignable locations.

But mostly it is used to distinguish functions that specify assigns clauses from other void functions and process them differently when dynamic frames are involved.

Reply:
I am not convinced that a `typedef` is the best way to address this problem. The number of permitted functions is very small, can't we just specifically check for those in each place that requires checking?

Reply:
Today we only provide 4 built-ins returning `__CPROVER_assignable_t`,
but the dynamic frames PRs builds on top of this to allow users to write their own functions returning `__CPROVER_assignable_t`. These user-defined functions can call other functions returning `__CPROVER_assignable_t` (built-in or user-defined), so the set is not necessarily small.

I intended  `__CPROVER_assignable_t` as an opaque type that represents a set of memory locations. It is defined as void in the front-end, and gets mapped to an actual representation by the instrumentation pass.

Reply:
My two cents, but I like the `__CPROVER_assignable_t` approach for representing sets of memory locations.